crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopInvariant.hs

```diff
@@ -0,0 +1,1151 @@
+            res_mem_impl <- storeMemJoinVariables
+              bak
+              header_mem_impl
+              mem_substitution
+              equality_substitution
+
+            -- == compute the list of "implicit parameters" that are relevant ==
+            let implicit_params = Set.toList $
+                  Set.difference
+                    (foldMap
+                       (\ (MapF.Pair _ e) ->
+                            -- filter out the special "noSatisfyingWrite" boolean constants
+                            -- that are generated as part of the LLVM memory model
+                            Set.filter ( \ (C.Some x) ->
+                                           not (List.isPrefixOf "cnoSatisfyingWrite"
+                                                (show $ W4.printSymExpr x))) $
+                            Set.map (\ (C.Some x) -> C.Some (W4.varExpr sym x)) $
+                              (W4.exprUninterpConstants sym (bodyValue e)))
+                       (MapF.toList (varSubst normal_substitution)))
+                    (Set.fromList (MapF.keys (varSubst normal_substitution)))
+
+            ?logMessage $ unlines $
+              ["Implicit parameters!"] ++
+              map (\ (C.Some x) -> show (W4.printSymExpr x)) implicit_params
+
+            -- == assert the loop invariant on the initial values ==
+
+            -- build a map where the current value is equal to the initial value
+            let init_state_map = MapF.map (\e -> e{ bodyValue = headerValue e })
+                                          (varSubst normal_substitution)
+            -- construct the loop invariant
+            initial_loop_invariant <- loop_invariant InitialInvariant implicit_params init_state_map
+            -- assert the loop invariant as an obligation
+            C.addProofObligation bak
+               $ C.LabeledPred initial_loop_invariant
+               $ C.SimError loc "initial loop invariant"
+
+            -- == assume the loop invariant on the arbitrary state ==
+
+            -- build a map where the current value is equal to the widening variable
+            let hyp_state_map = MapF.mapWithKey (\k e -> e{ bodyValue = k })
+                                                (varSubst normal_substitution)
+            -- construct the loop invariant to assume at the loop head
+            hypothetical_loop_invariant <- loop_invariant HypotheticalInvariant implicit_params hyp_state_map
+            -- assume the loop invariant
+            C.addAssumption bak
+              $ C.GenericAssumption loc "loop head invariant"
+                hypothetical_loop_invariant
+
+            -- == set up the state with arbitrary values to run the loop body ==
+
+            writeIORef fixpoint_state_ref $
+              CheckFixpoint
+                FixpointRecord
+                { fixpointBlockId = block_id
+                , fixpointAssumptionFrameIdentifier = undefined --assumption_frame_identifier
```

Comment:
Oh, good question.  I think it's safe because that assumption frame only exists when we are constructing the fixpoint, but I don't know if this is actually a good way to do this.